### PR TITLE
[FIXED] Check for invalid stream name in sources

### LINKED ISSUE #4141: JetStream Create API fails to reject some invalid source configurations
## Defect

An invalid stream configuration (missing names for some of the sources) is detected if using the NATS CLI to add the stream asset, but is not detected by the stream create API, e.g. when creating a stream with NACK or through a custom NATS program.

CLI detects:

```bash
$ nats --server "nats://UserB1:redacted@k8s.redacted.net:4222" str add 'BOGUS-SOURCES' --config ./bogus-sources.json
nats: error: could not create Stream: configuration validation failed: /sources/1/name: , /sources/1/name: length must be >= 1, but got 0, /sources/1/name: does not match pattern '^[^.*>]+$', /sources/2/name: , /sources/2/name: length must be >= 1, but got 0, /sources/2/name: does not match pattern '^[^.*>]+$'
```
Server allows:

```bash
$ cat ./bogus-sources.json | nats --server "nats://UserB1:redacted@k8s.redacted.net:4222" req '$JS.API.STREAM.CREATE.BOGUS-SOURCES' --replies 1
16:33:58 Reading payload from STDIN
16:33:58 Sending request on "$JS.API.STREAM.CREATE.BOGUS-SOURCES"
16:33:58 Received with rtt 8.268718ms
{"type":"io.nats.jetstream.api.v1.stream_create_response","config":{"name":"BOGUS-SOURCES","retention":"limits","max_consumers":-1,"max_msgs":-1,"max_bytes":-1,"max_age":0,"max_msgs_per_subject":-1,"max_msg_size":-1,"discard":"old","storage":"file","num_replicas":1,"duplicate_window":120000000000,"sources":[{"name":"LACR-STREAM"},{"name":"","external":{"api":"lacrJS.API","deliver":""}},{"name":"","external":{"api":"","deliver":"retail-lacr"}}],"allow_direct":false,"mirror_direct":false,"sealed":false,"deny_delete":false,"deny_purge":false,"allow_rollup_hdrs":false},"created":"2023-05-09T23:33:58.817532152Z","state":{"messages":0,"bytes":0,"first_seq":0,"first_ts":"0001-01-01T00:00:00Z","last_seq":0,"last_ts":"0001-01-01T00:00:00Z","consumer_count":0},"cluster":{"name":"nats","leader":"mybasic-nats-2"},"sources":[{"name":"LACR-STREAM","lag":0,"active":-1},{"name":"","external":{"api":"lacrJS.API","deliver":""},"lag":0,"active":-1},{"name":"","external":{"api":"","deliver":"retail-lacr"},"lag":0,"active":-1}],"did_create":true}
```
#### Versions of `nats-server` and affected client libraries used:

2.9.x (main)

#### Steps or code to reproduce the issue:

Create a stream configuration that has incomplete source definitions (here 2 of 3 sources missing name attribute):

```text
{
  "name": "BOGUS-SOURCES",
  "sources": [
    {
      "name": "LACR-STREAM"
    },
    {
      "external": {
        "api": "lacrJS.API"
      }
    },
    {
      "external": {
        "deliver": "retail-lacr"
      }
    }
  ],
  "retention": "limits",
  "max_consumers": -1,
  "max_msgs_per_subject": -1,
  "max_msgs": -1,
  "max_bytes": -1,
  "max_age": 0,
  "max_msg_size": -1,
  "storage": "file",
  "discard": "old",
  "num_replicas": 1,
  "duplicate_window": 120000000000,
  "sealed": false,
  "deny_delete": false,
  "deny_purge": false,
  "allow_rollup_hdrs": false,
  "allow_direct": false,
  "mirror_direct": false
}
```
Note: this is equivalent to a NACK stream asset YAML with two erroneous hyphens (externalApiPrefix, externalDeliverPrefix):

```yaml
apiVersion: jetstream.nats.io/v1beta2
kind: Stream
metadata:
  name: bogus-sources
spec:
  name: "BOGUS-SOURCES"
  sources:
    - name: "LACR-STREAM"
    - externalApiPrefix: "lacrJS.API"
    - externalDeliverPrefix: "retail-lacr"
  retention: limits
  maxConsumers: -1
  maxMsgsPerSubject: -1
  maxMsgs: -1
  maxBytes: -1
  maxAge: ""
  maxMsgSize: -1
  storage: file
  discard: old
  replicas: 1
  duplicateWindow: 120s
  denyDelete: false
  allowRollup: false
  allowDirect: false
```

#### Expected result:

NATS server will reject the stream creation request due to malformed request payload.

#### Actual result:

NATS server creates the asset, but attempts to interact with the stream through the NATS CLI later, e.g. `nats stream info BOGUS-SOURCES` fails with schema validation errors.

### PR DESCRIPTION
**Pull request by jnmoyne**
- [X] Link to issue, e.g. `Resolves #4141`
 - [X] Tests added
 - [X] Branch rebased on top of current dev
 - [X] Changes squashed to a single commit (described [here](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
 - [x] Build is green in Travis CI
 - [X] You have certified that the contribution is your original work and that you license the work to the project under the [Apache 2 license](https://github.com/nats-io/nats-server/blob/main/LICENSE)

Resolves #4141

### Changes proposed in this pull request:

Adds a check that the stream name of a stream source is valid and associated new error if it isn't.

### CODE REVIEW

File: server/errors.json [hunk 1]
```diff
@@ -1398,5 +1398,15 @@
     "help": "",
     "url": "",
     "deprecates": ""
+  },
```
Comment (derekcollison) [inline server/errors.json:1401]: Do we need a new one? Just use normal invalid stream name I think here.

**Review by derekcollison — COMMENTED** [review]
Rebase in top of dev should allow green build

File: server/errors.json (hunk 1 above)
Comment (jnmoyne) [reply server/errors.json:1401]: That's what I had done initially but then the error message the user gets is just "invalid stream name" but you don't know which stream name (the stream's name, the mirror name, one of the sources' name), so I added those two errors (invalid mirror and invalid source name). You prefer just a single one?

Comment (derekcollison) [reply server/errors.json:1401]: ok you raise a good point from user experience point of view.

**Comment by jnmoyne** [comment]
I just did a final adjustment of those two new errors but I believe it's done now

**Review by derekcollison — APPROVED** [review]
LGTM

**Pull request merged**